pkgs/swiftgen/swift2objc/lib/src/transformer/transformers/transform_referred_type.dart

```diff
@@ -0,0 +1,36 @@
+// Copyright (c) 2024, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:swift2objc/src/ast/_core/interfaces/declaration.dart';
+import 'package:swift2objc/src/ast/_core/shared/referred_type.dart';
+import 'package:swift2objc/src/transformer/transform.dart';
+
+DeclaredType transformReferredType(
+  ReferredType referredType,
+  TransformationMap transformationMap,
+) {
+  if (referredType is GenericType) {
+    throw UnimplementedError("Generic types are not supported yet");
+  }
+
+  referredType as DeclaredType;
+
+  if (referredType.isObjcRepresentable) return referredType;
+
+  if (transformationMap[referredType.declaration] != null) {
+    return transformationMap[referredType.declaration]!.asDeclaredType;
+  }
+
+  final transformedDeclaration = transformDeclaration(
+    referredType.declaration,
+    transformationMap,
+  );
+
+  transformationMap[referredType.declaration] = transformedDeclaration;
```

Comment:
Well the problem is that the insertion in this case will take place after the whole class has been transformed (including its methods). This will lead to an infinite loop in a situation like this:

```swift
class ClassA {
  func methodA(param: ClassB) {}
}

class ClassB {
  func methodB(param: ClassA) {}
}
```
- Say we start with transforming `ClassA`.
- We won't add it to the map until we transform `methodA` as well.
- `methodA` references `ClassB` so we have to transform it too.
- `methodB` references `ClassA` which we're supposedly already working on, but we don't know that cause we didn't add it to the map so we'll attempt transforming it again and enter a loop.

That's why I have add the class to the map before attempting to transform any of its members.

Reply:
Ok. I guess you could pass functions that do that stuff, but that's probably over-complicated. This is fine.